equivocation.go

```diff
@@ -0,0 +1,110 @@
+package f3
+
+import (
+	"bytes"
+	"slices"
+	"sync"
+
+	"github.com/filecoin-project/go-f3/gpbft"
+	"github.com/libp2p/go-libp2p/core/peer"
+)
+
+// zero value is valid
+type equivocationFilter struct {
+	lk              sync.Mutex
+	currentInstance uint64
+	// sentMessages map unique message slot to its signature
+	sentMessages  map[equivocationKey][]byte
+	activeSenders map[gpbft.ActorID][]peer.ID
+}
+
+type equivocationKey struct {
+	Sender gpbft.ActorID
+	Round  uint64
+	Phase  gpbft.Phase
+}
+
+func (ef *equivocationFilter) formKey(m gpbft.GMessage) equivocationKey {
+	return equivocationKey{
+		Sender: m.Sender,
+		Round:  m.Vote.Round,
+		Phase:  m.Vote.Phase,
+	}
+}
+
+func (ef *equivocationFilter) ProcessBroadcast(peerID peer.ID, m *gpbft.GMessage) bool {
+	ef.lk.Lock()
+	defer ef.lk.Unlock()
+
+	if m.Vote.Instance < ef.currentInstance {
+		// disallow past instances
+		log.Debugw("disallowing broadcast for past message", "sender", m.Sender, "instance",
+			m.Vote.Instance, "currentInstance", ef.currentInstance)
+		return false
+	}
+	// lazy initialization or moved onto new instance
+	if ef.sentMessages == nil || m.Vote.Instance > ef.currentInstance {
+		// bump to next instance
+		ef.currentInstance = m.Vote.Instance
+		ef.sentMessages = make(map[equivocationKey][]byte)
+		ef.activeSenders = make(map[gpbft.ActorID][]peer.ID)
+	}
+
+	key := ef.formKey(*m)
+	sig, ok := ef.sentMessages[key]
+	if ok {
+		log.Debugw("disallowing broadcast due to signature mismatch", "sender", m.Sender,
+			"instance", m.Vote.Instance, "round", m.Vote.Round, "phase", m.Vote.Phase)
+		return bytes.Equal(sig, m.Signature)
+	}
+
+	// save the signature
+	ef.sentMessages[key] = m.Signature
+
+	// save ourselves as one of the senders
+	senders := ef.activeSenders[m.Sender]
+	if !slices.Contains(senders, peerID) {
+		senders = append(senders, peerID)
+		slices.Sort(senders)
+		ef.activeSenders[m.Sender] = senders
+	}
+
+	// if there are multiple senders, only broadcast if we are the smallest one
+	return senders[0] == peerID
+}
+
+func (ef *equivocationFilter) ProcessReceive(peerID peer.ID, m *gpbft.GMessage) {
+	ef.lk.Lock()
+	defer ef.lk.Unlock()
+
+	if m.Vote.Instance != ef.currentInstance {
```

Comment:
This conflicts with gpbft logic: Messages from future instances [are queued](https://github.com/filecoin-project/go-f3/blob/a4deda5d9aaf80eb4c366dab56bf1987283c399f/gpbft/participant.go#L400). 

Ditto re focusing on equivocations only, otherwise we end up with diverging logic.

Reply:
ProcessBroadcast acceptes messages and deems them to be fine to broadcast or not
ProcessRecieve passively observes messages and tries to find cases where we equivocate messages we sent.

This function won't filter anything; it only observes.
The equivocation filter filters only on broadcasts.

Reply:
I see, thanks. Where would it be called from Kuba? Do you plan to integrate it in this PR?

Reply:
Yes, I plan to integrate it. It would be called either from late stage of pubsub validator or from message processing.